Sources/PenguinStructures/FixedSizeArray.swift

```diff
@@ -0,0 +1,269 @@
+      try body(
+          UnsafeMutableBufferPointer<Element>(
+              start: UnsafeMutableRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at the start.
+  func prepending(_ newElement: Element) -> ArrayN<Self> {
+    .init(head: newElement, tail: self)
+  }
+}
+
+/// A fixed sized collection of 0 elements.
+public struct Array0<T> : FixedSizeArray {
+  public init() {  }
+  
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.isEmpty`
+  @inline(__always)
+  public init<Source: Collection>(_ source: Source)
+      where Source.Element == Element
+  {
+    precondition(source.isEmpty, "Too many elements in source")
+  }
+  
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  public init(_ source: ArrayN<Self>, removingAt targetPosition: Index) {
+    precondition(targetPosition == 0, "Index out of range.")
+    self.init()
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at `targetPosition`.
+  public func inserting(_ newElement: Element, at i: Index) -> ArrayN<Self> {
+    precondition(i == 0, "Index out of range.")
+    return .init(head: newElement, tail: self)
+  }
+
+  // ======== Collection Requirements ============
+  public typealias Index = Int
+  
+  /// Accesses the element at `i`.
+  public subscript(i: Index) -> T {
+    get { fatalError("index out of range") }
+    set { fatalError("index out of range") }
+  }
+  
+  /// The position of the first element.
+  public var startIndex: Index { 0 }
+  
+  /// The position just past the last element.
+  public var endIndex: Index { 0 }
+}
+
+// ----- Standard conditional conformances. -------
+extension Array0 : Equatable where T : Equatable {}
+extension Array0 : Hashable where T : Hashable {}
+extension Array0 : Comparable where Element : Comparable {
+  public static func < (l: Self, r: Self) -> Bool { return false }
+}
+
+/// A fixed sized collection that stores one more element than `Tail` does.
+public struct ArrayN<Tail: FixedSizeArray> : FixedSizeArray {
+  private var head: Element
+  private var tail: Tail
+  
+  public typealias Element = Tail.Element
+
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.count == c`, where `c` is the capacity of instances.
+  @inline(__always)
+  public init<Source: Collection>(_ source: Source)
+      where Source.Element == Element
+  {
+    head = source.first!
+    tail = .init(source.dropFirst())
+  }
+
+  /// Creates an instance containing `head` followed by the contents of
+  /// `tail`.
+  // Could be private, but for a test that is using it.
```

Comment:
Do we need the doc comments adjacent to the declaration? Or is it okay if there's a regular comment in the way?

Edit: nvm, don't worry about this. When I re-gen docs next, I'll fix this up if required.

Reply:
I suppose it depends on your doc extraction tool.  The ones at Apple supported this arrangement.